modules/renter/contractor/host_integration_test.go

```diff
@@ -741,3 +742,78 @@ func TestResync(t *testing.T) {
 	}
 	editor.Close()
 }
+
+func TestDownloaderCaching(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	t.Parallel()
+	// create testing trio
+	h, c, _, err := newTestingTrio("TestDownloaderCaching")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+
+	// get the host's entry from the db
+	hostEntry, ok := c.hdb.Host(h.ExternalSettings().NetAddress)
+	if !ok {
+		t.Fatal("no entry for host in db")
+	}
+
+	// form a contract with the host
+	contract, err := c.managedNewContract(hostEntry, 10, c.blockHeight+100)
+	if err != nil {
+		t.Fatal(err)
+	}
+	c.mu.Lock()
+	c.contracts[contract.ID] = contract
+	c.mu.Unlock()
+
+	// create a downloader
+	d1, err := c.Downloader(contract.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// create another downloader
+	d2, err := c.Downloader(contract.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// downloaders should match
+	if d1 != d2 {
+		t.Fatal("downloader was not cached")
+	}
+
+	// close one of the downloaders; it should not fully close, since d1 is
+	// still using it
+	d2.Close()
+
+	// create another downloader
+	d3, err := c.Downloader(contract.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// downloaders should match
+	if d3 != d1 {
+		t.Fatal("closing one client should not fully close the downloader")
+	}
+
+	// close both downloaders
+	d1.Close()
+	d2.Close()
+
+	// create another downloader
+	d4, err := c.Downloader(contract.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// downloaders should match
+	if d4 == d1 {
+		t.Fatal("downloader should not have been cached after all clients were closed")
+	}
+}
```

Comment:
they don't do any downloading, but I guess I can check that they are not in `c.downloaders` after closing.